drivers/flash/Kconfig.mspi

```diff
@@ -4,7 +4,8 @@
 config FLASH_MSPI
 	bool "MSPI flash drivers"
 	default y
-	depends on FLASH
+	depends on DT_HAS_ZEPHYR_MSPI_EMUL_FLASH_ENABLED || \
+		   DT_HAS_MSPI_ATXP032_ENABLED
```

Comment:
I really dislike the `driver (devictree) > subsystem` dependency, perhaps introducing `config FLASH_MSPI_DRIVER_ENABLED` keeps it cleaner.

But why can't we make `FLASH_MSPI` an option that needs to be enabled explicitly?

Reply:
> I really dislike the `driver (devictree) > subsystem` dependency, perhaps introducing `config FLASH_MSPI_DRIVER_ENABLED` keeps it cleaner.
> 
> But why can't we make `FLASH_MSPI` an option that needs to be enabled explicitly?

I'd hate having all that DT_HAS_* macros there as it will look very messy as the list of drivers grow.
I was thinking if there is anyway to put a generated Kconfig there based on the common include binding. In my case, it is mspi-device.yaml. 
I don't think currently there is anyway to do that without making the DTS compatible string a bit more messy. i.e. "ambiq,mspi-device", "mspi-aps6404l", "mspi-device"; Then I could put a single DT_HAS_MSPI_DEVICE_ENABLED there. But I think it is kind of an overkill.

What do you mean by driver (devictree) > subsystem dependency? I thought this is supposedly the correct way. The CONFIG_FLASH=y is always explicit. FLASH_MSPI still depends on FLASH even if I deleted it there as the Kconfig.mspi is included within if FLASH in https://github.com/zephyrproject-rtos/zephyr/blob/4d0a8c1e8b8316ec5071a3fcc98cd292d3bc4e6d/drivers/flash/Kconfig#L186

Reply:
I would remove `default y` from `config FLASH_MSPI` and add these explicitly to boards that need it.

Reply:
Another option is to drop `config FLASH_MSPI` and let the drivers depend on `config FLASH` directly.

The `config FLASH_MSPI` does not impact the build whatsoever except allowing the two current drivers. Or would this change in the future? AFAICS they just implement flash APIs.

Reply:
Similar thing has been done in Kconfig.mcux. This allows us not to repeatly write common things for all its members i.e. FLASH_MSPI_EMUL_DEVICE and FLASH_MSPI_ATXP032. So I think keeping `config FLASH_MSPI` is much cleaner.
The `config FLASH_MSPI` currently generate cmake warning for non-compatible platforms. I'd say it is a temp fix until we could figure out how to generate a macro for common include bindings without flooding the output .config file.

Reply:
Having CONFIG_FLASH_MSPI across boards, examples and tests seems redundant. 

> Drivers on these busses should use select to allow the automatic activation of the necessary bus drivers when devices on the bus are enabled in the devicetree.
from https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations

Reply:
You left out the important part:
> An exemption are busses like I2C and SPI, and following the same thought process things like MFD as well.

And you select the bus as is documented with `select MSPI`.

> The config FLASH_MSPI currently generate cmake warning for non-compatible platforms. I'd say it is a temp fix until we could figure out how to generate a macro for common include bindings without flooding the output .config file.

Let's make it a hidden option then, that is selected by the drivers:

```
# Copyright (c) 2024 Ambiq Micro Inc. <www.ambiq.com>
# SPDX-License-Identifier: Apache-2.0

menu "MSPI flash driver"

config FLASH_MSPI
	bool
	select FLASH_HAS_DRIVER_ENABLED
	select MSPI
	help
	  MSPI flash drivers are enabled.

config FLASH_MSPI_EMUL_DEVICE
	bool "MSPI flash device emulator"
	default y
	depends on DT_HAS_ZEPHYR_MSPI_EMUL_FLASH_ENABLED
	select FLASH_MSPI
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_EXPLICIT_ERASE

config FLASH_MSPI_ATXP032
	bool "MSPI ATXP032 driver"
	default y
	depends on DT_HAS_MSPI_ATXP032_ENABLED
	select FLASH_MSPI
	select FLASH_HAS_PAGE_LAYOUT
	select FLASH_HAS_EXPLICIT_ERASE
	select FLASH_JESD216
	select MSPI_AMBIQ_AP3 if SOC_SERIES_APOLLO3X

endmenu

```

`depends on FLASH` is redundant because of `flash/Kconfig`

Reply:
Well that changed the hierarchy and looks a bit weird. It is not perfect either way.

Reply:
I'd say it's a common way of solving things using `Kconfig`, and avoids summing up all possible drivers.

Reply:
I have an idea for generating common includes in bindings. 
We could generate a compat DT_HAS macro for those that has a compatible string in yaml.
In my case, with ambiq,mspi-device or mspi-atxp032 includes mspi-device and with compatible: "mspi-device" added, we then generate DT_HAS_* macro whenever we encouter `compatible` in the includes as well.

Reply:
Sure, but why overcomplicate things with the need for users to know all devicetree compatibles/kconfig options?

The way I see this, is that these are just flash drivers, and if they are enabled in devicetree and `CONFIG_FLASH` is enabled, they should be compiled in. Having the additional option to disable them with `CONFIG_FLASH_MSPI=n` is more counterintuitive IMO.

Reply:
> Sure, but why overcomplicate things with the need for users to know all devicetree compatibles/kconfig options?
> 
> The way I see this, is that these are just flash drivers, and if they are enabled in devicetree and `CONFIG_FLASH` is enabled, they should be compiled in. Having the additional option to disable them with `CONFIG_FLASH_MSPI=n` is more counterintuitive IMO.

oops, my bad, I should delete the prompt for config FLASH_MSPI. It is my intention for it to be hidden. My goal was to have a hierachical structure and avoid having to write all the same things to its members.

Reply:
> oops, my bad, I should delete the prompt for config FLASH_MSPI. It is my intention for it to be hidden. My goal was to have a hierachical structure and avoid having to write all the same things to its members.

The more reason to let the symbol be selected by the driver options instead of constructing `default y` and `depends on`.

By the way, the same should be done for `MEMC_MSPI`.

Reply:
I see your point. 
Generating common include DT_HAS macro would still litter...
I will make the change.